#7773: ol Graticule layer support for different projections + print support

mbarto opened the pull request:
## Description
<!-- A few sentences describing the overall goals of the pull request' s commits. -->
Improvements to the graticule layer type:
 - new implementation for openlayers, compatible with the standard one, that delegates to MapGridsUtils most of the work, so that the logic is reusable for maps and printing
 - added support for grids in different projections (ol one only supports lat-lon grids)
 - added support for printing

As a bonus:
 - a plugin for printing is included, that allows adding a graticule to the printed map
 - various fixes / improvements to vector styling to support what is needed by the graticule

Some of the fixes need the latest mapfish-print print-lib.jar (from master) that contains the following fixes / improvements:
 - support for halo in labels / text (outlineLabelColor, outlineLabelWeight)
 - support for labels / text rotation
 - support for UTF-8 chars in labels / text


**Please check if the PR fulfills these requirements**
- [x] The commit message follows our guidelines: https://github.com/geosolutions-it/MapStore2/blob/master/CONTRIBUTING.md
- [x] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)


**What kind of change does this PR introduce?** (check one with "x", remove the others)
 - [ ] Bugfix
 - [x] Feature
 - [ ] Code style update (formatting, local variables)
 - [ ] Refactoring (no functional changes, no api changes)
 - [ ] Build related changes
 - [ ] CI related changes
 - [ ] Other... Please describe:

<!-- add here the ReadTheDocs link (if needed) -->

## Issue
#7773 

## Breaking change
**Does this PR introduce a breaking change?** (check one with "x", remove the other)
 - [ ] Yes, and I documented them in migration notes
 - [x] No

<!-- If this PR contains a breaking change, please describe the impact and migration path for existing applications -->

## Other useful information

offtherailz commented on web/client/utils/grids/GridLabelsUtils.js:
```diff
@@ -0,0 +1,128 @@
+ * @return {number} Modulo.
+ */
+function modulo(a, b) {
+    var r = a % b;
+    return r * b < 0 ? r + b : r;
```
((a % b ) + b ) % b should do the same, I think :smile:  No problem to do it more explicit. 
ref: 
https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Remainder

offtherailz reviewed (changes requested):
Hi @mbarto . Here my first comments. Sorry if I don't finish but I need more information to contiue reviewing it. 
- [ ] Please check and fix warning for linter camel case. in `Graticule.js`. 
- [ ] It is needed to enhance the legacy styler instead to enhancing the new style rules ?
- [ ] In order to test this I need also some fix on mapfish print? I ask you before starting testing and finalizing the review.  I see you are using several unicode chars  in code, and one of the point was that. 
- [ ] I see several usages of `detached` layer in openlayers and cesium implementation you did over the time, but I don't have a definition of that attribute. Can you explaing (and maybe document in a comment what it means, so the usage is clear for the future. 

I have to examinate and test the PR more in detail, but you can start fixing these and answering these questions, so I can better understand your work.

mbarto commented:
@offtherailz the fixes to mapfish-print are in the master branch of https://github.com/geosolutions-it/mapfish-print, and the corresponding jar has already been deployed to geosolutions maven repository.

offtherailz commented:
Great, so it should be already deployed on mapstore-dev, I will test on it while waiting for the other points. Thank you @mbarto for the info.

offtherailz commented:
@mbarto I added the points after testing in the review. Nice feature.

mbarto commented:
@offtherailz 

> Hi @mbarto . Here my first comments. Sorry if I don't finish but I need more information to contiue reviewing it.
> 
> * [ ]  Please check and fix warning for linter camel case. in `Graticule.js`.

Fixed

> * [ ]  It is needed to enhance the legacy styler instead to enhancing the new style rules ?

The rationale behind this choice is the following:  I tried to use the same path used for vector layers styling, that for openlayers uses VectorStyle.getStyle.
This one is already able to recognize the style type and use it properly (if you use a geostyle style, it should work); currently I see two different paths: either is a geostyler style, or a legacy style; annotations and drawing tools live on a different world, to my understanding, and are not integrated properly.

The story is a bit different for printing, where a specific set of transformers are used to build an ol2 style object (used by mapfish-print) from the vector style(s), and these transformers at the moment support only legacy style or annotation styles. This means that a geostyler style should not be usable, at the moment, for printing.

That said, I heard that an attempt to rationalize vector styling should be under way. In that case, you should consider having a general purpose converter (a la VectorStyle.getStyle) also for vector printing.

> * [x]  In order to test this I need also some fix on mapfish print? I ask you before starting testing and finalizing the review.  I see you are using several unicode chars  in code, and one of the point was that.
> * [ ]  I see several usages of `detached` layer in openlayers and cesium implementation you did over the time, but I don't have a definition of that attribute. Can you explaing (and maybe document in a comment what it means, so the usage is clear for the future.

I added some comments in the code. A detached layer is a layer that is not added/removed to the map via map.addLayer / map.removeLayer, so we need to work in different way with them. Both cesium and openlayers include some layers that need to be used this way.

Some refactoring on this part would probably make this concept more clear, I agree.

> 
> # Printing

> 
> * [ ]  In openlayers, I have an error:
>   Error:

Should be fixed now.
> 
> * [ ]  In cesium looks good. Maybe the default text is not easy to read (small), but acceptable.
>   ![image](https://user-images.githubusercontent.com/1279510/154734320-ebe2358b-14d8-4e36-910a-f879fbca5302.png)
> 

Take into account that the Cesium implementation is untouched and not part of the scope of this work.

> # Mix of the two
> I tried to print a map with both graticule turned on and layer graticule in the map. This is the final result. Not sure the color are the same, but I think this is not intended to be supported isn't it?

No, I don't think it make sense to have the Graticule both from the map and as an option in the printing panel.

offtherailz reviewed (approved):
Everything is clear and correct for me. 
We can merge. Thank you for your work.
